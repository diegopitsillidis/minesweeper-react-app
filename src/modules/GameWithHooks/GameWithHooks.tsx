import React, { FC, useState } from 'react' ;

import { GameLevels, LevelNames, GameSettings } from '@/modules/GameSettings';

import { Field, emptyFieldGenerator, CellState } from '@/helpers/Field';
import { Top } from '@/components/Top';
import { Scoreboard } from '@/components/Scoreboard';
import { Grid } from '@/components/Grid/Grid';
import { GameArea, Wrapper, GameOver } from '@/components/Game';

export const GameWithHooks: FC = () => {
    const [level, setLevel] = useState<LevelNames>('beginner');

    const [size, bombs] = GameSettings[level];

    const playerField = emptyFieldGenerator(size, CellState.hidden);

    return (
        <Wrapper>
            <Top feature='Flag' firstAction='Ctrl' secondAction='Click'>
                Minesweeper
            </Top>
            <GameArea>
                <Scoreboard
                    time='0'
                    bombs='10'
                    levels={GameLevels}
                    defaultLevel={level}
                    onChangeLevel={({ target: { value } }) => 
                        setLevel(value as LevelNames)
                    }
                    onReset={() => null}
                />
                <GameOver onClick={() => null} isWin={true} />
                <Grid onClick={() => null} onContextMenu={() => null}>
                    {playerField}
                </Grid>
            </GameArea>
        </Wrapper>
    );
};
