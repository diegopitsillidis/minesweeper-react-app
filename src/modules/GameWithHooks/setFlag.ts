import { detectSolvedPuzzle } from "@/core/detectSolvedPuzzle";
import { CellState, Coords, Field } from "@/core/Field";

/**
 * Set flat to the cell
 * @param {Coords} coords
 * @param {Field} playerField
 * @param {Field} gameField
 * @returns {[Field, boolean, number]}
 */
export const setFlag = (
    coords: Coords,
    playerField: Field,
    gameField: Field,
    prevFlagCounter: number,
    bombs: number
): [Field, boolean, number] => {
    console.log(gameField);
    const [y, x] = coords;
    const cell = playerField[y][x]

    const { flag, weakFlag, hidden } = CellState;

    switch (cell) {
        case flag:
            playerField[y][x] = weakFlag;
            break;
        case weakFlag:
            playerField[y][x] = hidden;
            break;
        case hidden:
            if(prevFlagCounter < bombs){
                playerField[y][x] = flag;
            }
            break;
    }
    const [isSolved, flagCounter] = detectSolvedPuzzle(playerField, gameField);

    return [playerField, isSolved, flagCounter];
}