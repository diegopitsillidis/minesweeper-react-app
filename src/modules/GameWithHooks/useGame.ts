import { Field, generateFieldWithDefaultState, CellState, fieldGenerator, Coords } from '@/core/Field';
import { useState } from 'react';
import { LevelNames, GameSettings } from '../GameSettings';
import { openCell } from '@/core/openCell';
import { setFlag } from './setFlag';
import { useTime } from './useTime';

interface ReturnType {
    level: LevelNames;
    time: number;
    isGameOver: boolean;
    isGameStarted: boolean;
    isWin: boolean;
    settings: [number, number];
    playerField: Field;
    gameField: Field;
    flagCounter: number;
    onClick: (coords: Coords) => void;
    onContextMenu: (coords: Coords) => void;
    onChangeLevel: (levl: LevelNames) => void;
    onReset: () => void;
}

export const useGame = (): ReturnType => {
    const [level, setLevel] = useState<LevelNames>('beginner');
    
    const [isGameOver, setIsGameOver] = useState(false);
    const [isWin, setIsWin] = useState(false);
    const [isGameStarted, setIsGameStarted] = useState(false);

    const [time, resetTime] = useTime(isGameStarted, isGameOver);

    const [flagCounter, setFlagCounter] = useState(0);

    const setGameOver = (isSolved = false) => {
        setIsGameOver(true);
        setIsWin(isSolved);
    };

    const [size, bombs] = GameSettings[level];

    const [playerField, setPlayerField] = useState<Field>(
        generateFieldWithDefaultState(size, CellState.hidden)
    );

    const [gameField, setGameField] = useState<Field>(
        fieldGenerator(size, bombs / (size*size))
    );

    const onClick = (coords: Coords)=> {
        if (!isGameStarted) {
            setIsGameStarted(true);
        }

        try{
            const [newPlayerField, isSolved] = openCell(
                coords, 
                playerField, 
                gameField,
            );
            if (isSolved) {
                setGameOver(isSolved);
            }
            setPlayerField([...newPlayerField]);
        } catch (e) {
            console.log(e);
            setPlayerField([... gameField]);
            setGameOver();
        }
        
    }

    const onContextMenu = (coords: Coords) => {
        if (!isGameStarted) {
            setIsGameStarted(true);
        }
        
        const [newPlayerField, isSolved, newFlagCounter] = setFlag(
            coords, 
            playerField, 
            gameField,
            flagCounter,
            bombs
        );
        setFlagCounter(newFlagCounter);

        if (isSolved) {
            setGameOver(isSolved);
        }
        setPlayerField([...newPlayerField]);
    };

    const resetHandler = ([size, bombs]: [number, number]) => {
        const newGameField = fieldGenerator(size, bombs / (size * size))
        const newPlayerField = generateFieldWithDefaultState(
            size, 
            CellState.hidden
        );

        setGameField([...newGameField])
        setPlayerField([...newPlayerField]);
        setIsGameOver(false);
        setIsWin(false);
        setIsGameStarted(false);
        resetTime();
        setFlagCounter(0);
    }

    const onChangeLevel = (level: LevelNames) => {
        setLevel(level);
        const newSettings = GameSettings[level];
        resetHandler(newSettings);
    };

    const onReset = () => resetHandler([size, bombs])

    return {
        level,
        time,
        isGameOver,
        isGameStarted,
        isWin,
        settings: [size, bombs],
        playerField,
        gameField,
        flagCounter,
        onClick,
        onContextMenu,
        onChangeLevel,
        onReset,
    };
};