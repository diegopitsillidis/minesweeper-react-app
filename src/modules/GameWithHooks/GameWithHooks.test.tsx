import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import { GameWithHooks } from './GameWithHooks';

describe('GameWithHooks test cases', () => {
    describe('Render behaiviour', () => {
        it('Render game field by default', () => {
            const { asFragment } = render(<GameWithHooks />);

            expect(screen.getAllByRole('cell')).toHaveLength(81);  

            expect(asFragment()).toMatchSnapshot();
        });
        it('onChange game level handler', () => {
            render(<GameWithHooks />);

            expect(screen.getAllByRole('cell')).toHaveLength(81);  

            userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');

            expect(screen.getAllByRole('cell')).toHaveLength(256);

            userEvent.selectOptions(screen.getByRole('combobox'), 'expert');
            
            expect(screen.getAllByRole('cell')).toHaveLength(484);
        });
    })
});