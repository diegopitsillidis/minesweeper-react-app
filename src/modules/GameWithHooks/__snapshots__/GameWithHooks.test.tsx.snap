// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render behaiviour Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1ct4owg"
  >
    <header
      class="css-1mm4p1j"
    >
      <h1
        class="css-aqrpco"
      >
        Minesweeper
      </h1>
      <legend
        class="css-y5wpvr"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-1th0bce"
        >
          <span
            class="css-1eygtyr"
          >
            Ctrl
          </span>
           +  
          <span
            class="css-1r5wa4u"
          >
            Click
          </span>
        </code>
      </legend>
    </header>
    <div
      class="css-1bh93n8"
    >
      <div
        class="css-1x0xxpv"
      >
        <div
          class="css-10qcrn4"
        >
          0
        </div>
        <select
          class="css-cvsgr"
        >
          <option
            class="css-1ibfzol"
            value="beginner"
          >
            beginner
          </option>
          <option
            class="css-1ibfzol"
            value="intermediate"
          >
            intermediate
          </option>
          <option
            class="css-1ibfzol"
            value="expert"
          >
            expert
          </option>
        </select>
        <button
          class="css-2jpmyz"
        >
          ðŸ˜Š
        </button>
        <div
          class="css-10qcrn4"
        >
          10
        </div>
      </div>
      <div
        class="css-1p4hvro"
      >
        ðŸ˜Ž
      </div>
      <div
        class="css-1a9cl3f"
        size="9"
      >
        <div
          class="css-z6puck"
          data-testid="10_0,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_0,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_1,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_2,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_3,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_4,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_5,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_6,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_7,8"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,0"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,1"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,2"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,3"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,4"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,5"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,6"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,7"
          role="cell"
        />
        <div
          class="css-z6puck"
          data-testid="10_8,8"
          role="cell"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
